File: Easy/equivalent-resistance-circuit-building/equivalent-resistance-circuit-building.py
n = int(input())
r = {}

#Resistor names and values
for i in range(n):
    x=input().split()
    r[x[0]]=int(x[1])


c = input()

def findElements(resString):
    elements = []
    paranthesis = []
    hold = ""

    for character in resString:

        hold += character

        if character == " ":
            #If paranthesis' are matched, append hold without last space to elements
            if len(paranthesis) == 0:
                elements.append(hold.rstrip())
                hold = ""        

        elif character == "(" or character == "[":
            paranthesis.append(character)

        elif character == ")" or character == "]":
            paranthesis.pop(-1)        

    if hold != "":
        elements.append(hold)

    return elements

def findRes(s):
    if s in r:
        return r[s]
    else:
        #elements = re.findall(r"\[.*\]|\(.*\)|\w+",s[2:-2])
        elements = findElements(s[2:-2])
        #Series
        if s[0] == "(":
            return sum([findRes(element) for element in elements])
        
        #Parallel
        else:
            return 1/sum([1/findRes(element) for element in elements])
        
        

print(f"{findRes(c):.1f}")

File: Easy/equivalent-resistance-circuit-building/equivalent-resistance-circuit-building.js
const N = parseInt(readline());
const resistors = {};

//Resistor names and values
for (let i = 0; i < N; i++) {
    var inputs = readline().split(' ');
    const name = inputs[0];
    const R = parseInt(inputs[1]);
    resistors[name] = R;
}
const circuit = readline();

const findElements = (resString) => {
    let elements = [];
    let paranthesis = [];
    let hold = "";

    for(const character of resString){

        hold += character;

        if(character === " "){

            //If paranthesis' are matched, append hold without last space to elements
            if(paranthesis.length === 0){
                elements.push(hold.slice(0, -1));
                hold = "";
            }
        }
        else if(character === "(" || character === "["){
            paranthesis.push(character);
        }
        else if(character === ")" || character === "]"){
            paranthesis.pop();
        }
    }

    if(hold !== ""){
        elements.push(hold);
    }
    return elements;
};

const findRes = (resString) => {
    
    if(resString in resistors){
        return resistors[resString];
    }

    else{
        const elements = findElements(resString.slice(2,-2));

        //Series Connection
        if(resString[0] === "("){
            return elements.map(element => findRes(element)).reduce((element1,element2) => element1 + element2);
        }
        //Parallel Connection
        else{
            return 1/elements.map(element => 1/findRes(element)).reduce((element1,element2) => element1 + element2);
        }
    }
};

console.log(findRes(circuit).toFixed(1)); 
